File: nglpy_cuda/Graph.py
"""
    The API for using NGLPy with CUDA
"""
import nglpy_cuda as ngl
import numpy as np

from .utils import f32, i32
from .SKLSearchIndex import SKLSearchIndex

from threading import Thread
import sys
import os
import abc
if sys.version_info.major >= 3:
    from queue import Queue, Empty
    ABC = abc.ABC
else:
    from Queue import Queue, Empty
    ABC = abc.ABCMeta('ABC', (), {})


class Graph(ABC):
    """ A neighborhood graph that represents the connectivity of a given
    data matrix.

    Attributes:
        None
    """

    def __init__(self,
                 index=None,
                 max_neighbors=-1,
                 query_size=None,
                 cached=False):
        """Initialization of the graph object. This will convert all of
        the passed in parameters into parameters the C++ implementation
        of NGL can understand and then issue an external call to that
        library.

        Args:
            index (string): A nearest neighbor index structure which can
                be queried and pruned
            max_neighbors (int): The maximum number of neighbors to
                associate with any single point in the dataset.
            query_size (int): The number of points to process with each
                call to the GPU, this should be computed based on
                available resources
            cached (bool): Flag denoting whether the resulting computation
                should be stored for future use, will consume more memory,
                but can drastically reduce the runtime for subsequent
                iterations through the data.
        """
        self.X = np.array([[]], dtype=f32)
        self.max_neighbors = max_neighbors
        if index is None:
            self.nn_index = SKLSearchIndex()
        else:
            self.nn_index = index
        self.query_size = query_size
        self.cached = cached
        self.edges = None

    def build(self, X):
        """ Build a graph based on the incoming data

        Args:
            X (matrix): The data matrix for which we will be determining
                connectivity.
        """
        self.X = np.array(X, dtype=f32)
        N = len(X)

        if self.max_neighbors < 0:
            self.max_neighbors = min(1000, N)

        self.nn_index.fit(self.X)

        if self.query_size is None:
            self.query_size = self.compute_query_size()

        self.chunked = self.X.shape[0] > self.query_size

        # print('Problem Size: {}'.format(N))
        # print('  Query Size: {}'.format(self.query_size))
        # print('  GPU Memory: {}'.format(available_gpu_memory))
        # print('     Chunked: {}'.format(self.chunked))
        sys.stdout.flush()

        self.edge_list = Queue(self.query_size*10)
        self.needs_reset = False

        self.worker_thread = Thread(target=self.populate)
        self.worker_thread.daemon = True
        self.worker_thread.start()

    @abc.abstractmethod
    def compute_query_size(self):
        pass

    @abc.abstractmethod
    def collect_additional_indices(self, edges, indices):
        pass

    @abc.abstractmethod
    def prune(self, X, edges, indices=None, count=None):
        pass

    def populate_chunk(self, start_index):
        end_index = min(start_index+self.query_size, self.X.shape[0])
        count = end_index - start_index
        working_set = np.array(range(start_index, end_index))

        edges = self.nn_index.search(working_set, self.max_neighbors, False)

        indices = self.collect_additional_indices(edges, working_set)
        indices = np.array(indices, dtype=i32)
        X = self.X[indices, :]

        edges = self.prune(X, edges, indices, count)

        # We will cache these for later use
        if self.cached:
            self.edges[start_index:end_index, :] = edges[:count]

        # Since, we are taking a lot of time to generate these, then we
        # should give the user something to process in the meantime, so
        # don't remove these lines and make sure to return in the main
        # populate before the same process is done again.
        self.push_edges(edges[:count], indices[:count])

    def populate_whole(self):
        count = self.X.shape[0]
        working_set = np.array(range(count))
        edges = self.nn_index.search(working_set, self.max_neighbors, False)

        edges = self.prune(self.X, edges)

        if self.cached:
            self.edges = edges
        self.push_edges(edges)

    def populate(self):
        point_count = self.X.shape[0]
        start_index = 0
        if self.edges is not None:
            start_index = point_count

        fname = 'nglpy.checkpoint'
        if self.cached and os.path.isfile(fname):
            with open(fname, 'r') as f:
                start_index = int(f.read())

        if self.edges is None or start_index < point_count:
            data_shape = (point_count, self.max_neighbors)
            if self.cached:
                self.edges = np.memmap(
                    'edges.npy', dtype=i32, mode='w+', shape=data_shape)

            if self.chunked:
                while start_index < point_count:
                    self.populate_chunk(start_index)
                    start_index += self.query_size
                    start_index = min(start_index, point_count)
                    with open(fname, 'w') as f:
                        f.write(str(start_index))
                    print('Checkpoint: {}'.format(start_index))
            else:
                self.populate_whole()
        else:
            self.push_edges(self.edges)

    def push_edges(self, edges, indices=None):
        if indices is not None:
            valid_edges = ngl.get_edge_list(edges, indices)
        else:
            valid_edges = ngl.get_edge_list(edges)
        for edge in valid_edges:
            self.edge_list.put(edge)

    def restart_iteration(self):
        if not self.worker_thread.is_alive() and self.needs_reset:
            self.edge_list.queue.clear()
            self.needs_reset = False
            self.worker_thread = Thread(target=self.populate)
            self.worker_thread.daemon = True
            self.worker_thread.start()

    def __iter__(self):
        return self

    def next(self):
        return self.__next__()

    def __next__(self):
        if self.needs_reset:
            self.restart_iteration()
        while not self.edge_list.empty() or self.worker_thread.is_alive():
            try:
                next_edge = self.edge_list.get(timeout=1)
                return next_edge
            except Empty:
                pass
        self.needs_reset = True
        raise StopIteration

    def full_graph(self):
        neighborhoods = {}
        for (p, q, d) in self:
            p = int(p)
            q = int(q)
            if p not in neighborhoods:
                neighborhoods[p] = set()
            if q not in neighborhoods:
                neighborhoods[q] = set()
            neighborhoods[p].add(q)
            neighborhoods[q].add(p)
        return neighborhoods

    def neighbors(self, i):
        nn = []
        if self.edges is None:
            raise NotImplementedError

        for value in self.edges[i]:
            if value != -1:
                nn.append(value)
        return nn

File: nglpy_cuda/Graph.cpp
/*
 * Graph.h
 *
 *  Created on: Oct 22, 2018
 *      Author: maljovec
 */

#include "Graph.h"
#include "SearchIndex.h"
#include "ANNSearchIndex.h"
#include "ngl_cuda.h"
#include <vector>
#include <set>

Graph::Graph(SearchIndex *index,
             int maxNeighbors,
             bool relaxed,
             float beta,
             float p,
             int discreteSteps,
             int querySize)
    : mMaxNeighbors(maxNeighbors),
      mRelaxed(relaxed),
      mBeta(beta),
      mLp(p),
      mDiscreteSteps(discreteSteps),
      mQuerySize(querySize)
{
    if (index != NULL)
    {
        mSearchIndex = index;
    }
    else
    {
        mSearchIndex = new ANNSearchIndex(maxNeighbors, 0);
        mSelfConstructedIndex = true;
    }
}

void Graph::build(float *X, int N, int D)
{
    mData = X;
    mCount = N;
    mDim = D;
    mSearchIndex->fit(mData, mCount, mDim);

    size_t availableGPUMemory = nglcu::get_available_device_memory();
    if (mQuerySize < 0)
    {
        // Because we are using f32 and i32:
        int bytesPerNumber = 4;
        int k = maxNeighbors;
        int worstCase;

        // Worst-case upper bound limit of the number of points
        // needed in a single query
        if (mRelaxed)
        {
            // For the relaxed algorithm we need n*D storage for the
            // point locations plus n*k for the representation of the
            // input edges plus another n*k for the output edges
            // We could potentially only need one set of edges for
            // this version
            worstCase = D + 2 * k;
        }
        else
        {
            // For the strict algorithm, if we are processing n
            // points at a time, we need the point locations of all
            // of their neigbhors, thus in the worst case, we need
            // n + n*k point locations and rows of the edge matrix.
            // the 2*k again represents the need for two versions of
            // the edge matrix. Here, we definitely need an input and
            // an output array
            worstCase = (D + 2 * k) * (k + 1);
        }

        // If we are using the discrete algorithm, remember we need
        // to add the template's storage as well to the GPU
        if (mDiscreteSteps > 0)
        {
            availableGPUMemory -= mDiscreteSteps * bytesPerNumber;
        }

        int divisor = bytesPerNumber * worstCase;
        mQuerySize = std::min(availableGPUMemory / divisor, mCount);
    }
    mChunked = mQuerySize < mCount;
    populate();
}

void Graph::populate()
{
    if (mChunked)
    {
        mEdges = new int[mQuerySize * mMaxNeighbors];
        mDistances = new float[mQuerySize * mMaxNeighbors];
        populate_chunk(0);
    }
    else
    {
        mEdges = new int[mCount * mMaxNeighbors];
        mDistances = new float[mCount * mMaxNeighbors];
        populate_whole();
    }
}

void Graph::populate_chunk(int startIndex)
{
    mRowOffset = startIndex;
    int count = std::min(mCount - startIndex, mQuerySize);
    int edgeCount = count;
    int endIndex = startIndex + count;
    mSearchIndex->search(mRowOffset, count, mMaxNeighbors, mEdges, mDistances);

    std::set<int> additionalIndices;
    for (int i = 0; i < count; i++)
    {
        for (int k = 0; k < mMaxNeighbors; k++)
        {
            if (mEdges[i * mMaxNeighbors + k] < startIndex || mEdges[i * mMaxNeighbors + k] >= endIndex)
            {
                additionalIndices.insert(mEdges[i * mMaxNeighbors + k]);
            }
        }
    }

    if (additionalIndices.size() > 0)
    {
        std::vector<int> indices;
        for (int i = startIndex; i < endIndex; i++)
        {
            indices.push_back(i);
        }
        int extraCount = additionalIndices.size();
        for (auto it = additionalIndices.begin(); it != additionalIndices.end(); it++)
        {
            indices.push_back(*it);
        }
        if (!mRelaxed)
        {
            int *extraEdges = new int[extraCount * mMaxNeighbors];
            int *extraIndices = indices.data() + count;
            mSearchIndex->search(extraIndices, extraCount, mMaxNeighbors, extraEdges, NULL);
            delete extraIndices;
            edgeCount = count + extraCount;
            int *allEdges = new int[edgeCount * mMaxNeighbors];
            for (int i = 0; i < count; i++)
            {
                for (int k = 0; k < mMaxNeighbors; k++)
                {
                    allEdges[i * mMaxNeighbors + k] = mEdges[i * mMaxNeighbors + k];
                }
            }
            std::set<int> uniqueIndices(indices.begin(), indices.end());
            additionalIndices.clear();
            for (int i = 0; i < extraCount; i++)
            {
                for (int k = 0; k < mMaxNeighbors; k++)
                {
                    allEdges[(i + count) * mMaxNeighbors + k] = extraEdges[i * mMaxNeighbors + k];
                    additionalIndices.insert(extraEdges[i * mMaxNeighbors + k]);
                }
            }
            delete mEdges;
            delete extraEdges;
            mEdges = allEdges;

            for (auto it = additionalIndices.begin(); it != additionalIndices.end(); it++)
            {
                indices.push_back(*it);
            }
        }
    }

    float *X = new float[indices.size() * mDim];
    for (int i = 0; i < indices.size(); i++)
    {
        for (int d = 0; d < mDim; d++)
        {
            X[i * mDim + d] = mData[indices[i] * mDim + d];
        }
    }

    if (mDiscreteSteps > 0)
    {
        nglcu::prune_discrete(X, mEdges, indices.data(), count, mDim, edgeCount,
                              mMaxNeighbors, mRelaxed, mBeta, mLp);
    }
    else
    {
        nglcu::prune(X, mEdges, indices.data(), count, mDim, edgeCount,
                     mMaxNeighbors, mRelaxed, mBeta, mLp);
    }

    delete X;
}

void Graph::populate_whole()
{
    mSearchIndex->search(0, mCount, mMaxNeighbors, mEdges, mDistances);
    nglcu::prune(mData, mEdges, NULL, mCount, mDim, mCount,
                 mMaxNeighbors, mRelaxed, mBeta, mLp);
}

void Graph::restart_iteration()
{
    mCurrentCol = 0;
    mCurrentRow = 0;
    // If we have changed rows, let's ensure we don't need to run
    // another query
    if (chunked && mRowOffset > mCurrentRow)
    {
        populate_chunk(mCurrentRow);
    }
}

Edge Graph::next()
{
    Edge e;
    if(mIterationFinished) {
        // Set up the next round of iteration
        mIterationFinished = false;
        e.indices[0] = -1;
        e.indices[1] = -1;
        e.length = 0;
        return e;
    }
    int currentIndex = (mCurrentRow - mRowOffset) * mMaxNeighbors + mCurrentCol;
    e.distance = mDistances[currentIndex];
    if (mReversed)
    {
        e.indices[0] = mEdges[currentIndex];
        e.indices[1] = mCurrentRow;
    }
    else
    {
        e.indices[0] = mCurrentRow;
        e.indices[1] = mEdges[currentIndex];
    }

    mReversed = !mReversed;
    if (!mReversed)
    {
        advanceIteration();
        while(mEdges[(mCurrentRow - mRowOffset) * mMaxNeighbors + mCurrentCol] == -1) {
            advanceIteration();
        }
    }

    return e;
}

void Graph::advanceIteration() {
    mCurrentCol++;
    if (mCurrentCol >= mMaxNeighbors)
    {
        mCurrentCol = 0;
        mCurrentRow++;
        if (mCurrentRow >= mCount)
        {
            mCurrentRow = 0;
            mIterationFinished = true;
        }
        // If we have changed rows, let's ensure we don't need to run
        // another query
        if (chunked) {
            if (mCurrentRow - mRowOffset >= mQuerySize || mCurrentRow == 0)
            {
                populate_chunk(mCurrentRow);
            }
        }
    }
}

virtual Graph::~Graph()
{
    if (mSelfConstructedIndex)
    {
        delete mSearchIndex;
    }
    delete mEdges;
    delete mDistances;
}